units/host_allocator.go

```diff
@@ -202,10 +202,20 @@ func (j *hostAllocatorJob) Run(ctx context.Context) {
 	// and the hosts allocated for them,
 	// how long will it take the current fleet of hosts, plus the ones we spawned, to chew through
 	// the scheduled tasks in the queue?
+
 	scheduledDuration := distroQueueInfo.ExpectedDuration - distroQueueInfo.DurationOverThreshold
 
+	var totalOverdueInTaskGroups, CountDurationOverThresholdInTaskGroups int
+	for _, info := range distroQueueInfo.TaskGroupInfos {
+		if info.Name != "" {
+			totalOverdueInTaskGroups += info.CountWaitOverThreshold
+			CountDurationOverThresholdInTaskGroups += info.CountDurationOverThreshold
+		}
+	}
+	durationOverThreshNoTaskGroups := distroQueueInfo.CountDurationOverThreshold - CountDurationOverThresholdInTaskGroups
+	hostsAvail := nHostsFree + len(hostsSpawned) - durationOverThreshNoTaskGroups
```

Comment:
I don't understand: doesn't `hostsSpawned` include hosts spawned for tasks in task groups? Also, `scheduledDuration` includes task group tasks, doesn't it?

Reply:
I think nHostsFree and scheduledDuration can and should be corrected as you suggest. Thank you! It might not be possible to do this accurately all the time with hostsSpawned, because there's no strict accounting of spawned hosts by task groups and the number of spawned hosts might be different from the requested hosts in some cases. However, we already have versions of these metrics without taking spawned hosts into account, so this should be OK.